cockatrice/src/player.cpp

```diff
@@ -1126,46 +1126,175 @@ void Player::actCreateRelatedCard()
-    const QString &actionDisplayName = action->text();
-    createCard(sourceCard, dbNameFromTokenDisplayName(actionDisplayName));
+    //If there is a better way of passing a CardRelation through a QAction, please add it here.
+    QList<CardRelation *> relatedCards = QList<CardRelation *>();
+    relatedCards.append(sourceCard->getInfo()->getRelatedCards());
+    relatedCards.append(sourceCard->getInfo()->getReverseRelatedCards2Me());
+    CardRelation * cardRelation = relatedCards.at(action->data().toInt());
+
+    const QString &displayName = cardRelation->getName();
+    QString dbName = dbNameFromTokenDisplayName(displayName);
+    if(cardRelation->getIsVariable()) {
+        bool ok;
+        dialogSemaphore = true;
+        int count = QInputDialog::getInt(0, tr("Create tokens"), tr("Number:"), cardRelation->getDefaultCount(), 1, MAX_TOKENS_PER_DIALOG, 1, &ok);
+        dialogSemaphore = false;
+        if (!ok)
+            return;
+        for (int i = 0; i < count; i++) {
+            createCard(sourceCard, dbName);
+        }
+    } else if(cardRelation->getDefaultCount() > 1) {
+        for (int i = 0; i < cardRelation->getDefaultCount(); i++) {
+            createCard(sourceCard, dbName);
+        }
+    } else {
+        if (cardRelation->getDoesAttach()) {
+            createAttachedCard(sourceCard, dbName);
+        } else {
+            createCard(sourceCard, dbName);
+        }
+    }
 
    /*
     * If we made a token via "Token: TokenName"
     * then lets allow it to be created via create another
     */
-    CardInfo *cardInfo = db->getCard(dbNameFromTokenDisplayName(actionDisplayName));
-    setLastToken(cardInfo);
+    if (cardRelation->getCanCreateAnother()) {
+        CardInfo *cardInfo = db->getCard(dbNameFromTokenDisplayName(displayName));
+        setLastToken(cardInfo);
+    }
 }
 
+/**
+ *
+ */
 void Player::actCreateAllRelatedCards()
 {
     CardItem * sourceCard = game->getActiveCard();
     if(!sourceCard)
         return;
 
-    QStringList relatedCards = * new QStringList();
+    QList<CardRelation *> relatedCards = QList<CardRelation *>();
     relatedCards.append(sourceCard->getInfo()->getRelatedCards());
     relatedCards.append(sourceCard->getInfo()->getReverseRelatedCards2Me());
 
-    foreach (const QString &tokenName, relatedCards)
-    {
-        createCard(sourceCard, dbNameFromTokenDisplayName(tokenName));
+    QList<CardRelation *> nonExcludedRelatedCards = QList<CardRelation *>();
+    QString dbName;
+    CardRelation * cardRelation = nullptr;
+    int tokensTypesCreated = 0;
+
+    switch(relatedCards.length()) { //Is an if/elseif/else pattern better?
+        case 0: //if (relatedCards.length() == 0)
+            return;
+        case 1: //else if (relatedCards.length() == 1)
+            cardRelation = relatedCards.at(0);
+            tokensTypesCreated++;
+            dbName = dbNameFromTokenDisplayName(cardRelation->getName());
+            if(cardRelation->getIsVariable()) {
```

Comment:
The following ~20 lines of code are almost duplicated here, in lines 1228-1248 in this same method an in `Player::actCreateRelatedCard()`. It would be nice to put them inside a new function and just call it.